File: adapters/playwright/src/sessions.ts
// SPDX-License-Identifier: Apache-2.0
//
// Session manager for the Playwright adapter.
//
// Owns the lazy `Browser` launch, the per-session `BrowserContext`
// + `Page` allocation, and the per-session `ElementRegistry` that
// backs `Query` and `Extract`. R4.3 externalises session metadata
// to Redis (ADR-0023 §4 + §5): `register` writes the metadata,
// `validate` reads it on every non-Initialize RPC, `closeSession`
// deletes the key. The runtime objects (`Browser`, `BrowserContext`,
// `Page`, ElementRegistry entries) stay process-local — the §5
// restart-invalidation contract makes Pod restart equivalent to
// session loss for clients.
//
// Contract (R4.3):
//
//   - `register(sessionId)` — `Initialize` calls this. Writes a
//     `SessionMetadata` document to Redis with the current
//     `adapter_instance_id` and adds the id to the local
//     registered set. Throws if Redis is unreachable; the caller
//     maps the failure to gRPC `UNAVAILABLE`.
//   - `validate(sessionId)` — every non-Initialize RPC calls this
//     before doing any work. Returns one of three kinds:
//       * `ok` — Redis has the session and the stored
//         `adapter_instance_id` matches; TTL refreshed and
//         `last_active_at` updated.
//       * `unknown` — Redis has no entry for this session_id
//         (never created or TTL-expired). Caller maps to
//         `INVALID_ARGUMENT` per ADR-0009 §2.
//       * `different-instance` — Redis has the session but it
//         belongs to a different adapter instance (the §5
//         restart-invalidation case). Caller raises gRPC
//         `UNAVAILABLE` with the message "session belongs to a
//         different adapter instance; client must re-Initialize".
//   - `has(sessionId)` — true once `register` has been called for
//     the id locally and `closeSession`/`closeAll` have not yet
//     evicted it. Used for cheap membership checks where Redis
//     is not authoritative (the `closeSession` early-out path).
//   - `getOrCreatePage(sessionId)` — first call for a registered
//     id launches the shared `Browser` (if not already) and
//     creates a fresh `BrowserContext` + `Page`. Subsequent calls
//     with the same id return the same `Page`.
//   - `pageOf(sessionId)` — returns the live `Page` for a session
//     if one exists, or null if the session is registered but has
//     not yet navigated.
//   - `bumpGeneration` / `currentGeneration` / `allocateRefs` /
//     `lookupRef` — element-registry delegation, unchanged from
//     PR5/ADR-0010.
//   - `closeSession(sessionId)` — deletes the Redis key
//     (best-effort), evicts the local registration, forgets all
//     ElementRefs, and closes the per-session `BrowserContext`.
//   - `closeAll()` — closes every page, every context, the shared
//     `Browser`, and clears every local registration. Idempotent.
//     Does not enumerate Redis keys for deletion — restart
//     invalidation handles abandoned keys via TTL expiry.
//
// The class is constructed with a `BrowserFactory`, a
// `RedisClient`, and the adapter's `instanceId`. Unit tests pass
// an `ioredis-mock` instance and any `instanceId` string.

import type { Browser, BrowserContext, Locator, Page } from "playwright";

import { ElementRegistry, type RefLookup } from "./elements.js";
import {
  ADAPTER_NAME,
  type RedisClientLike,
  type SessionMetadata,
} from "./redis.js";

export type BrowserFactory = () => Promise<Browser>;

interface Session {
  context: BrowserContext;
  page: Page;
}

export class UnknownSessionError extends Error {
  readonly sessionId: string;
  constructor(sessionId: string) {
    super(`unknown session_id: ${JSON.stringify(sessionId)}`);
    this.name = "UnknownSessionError";
    this.sessionId = sessionId;
  }
}

export type ValidationResult =
  | { kind: "ok"; metadata: SessionMetadata }
  | { kind: "unknown" }
  | { kind: "different-instance"; storedInstanceId: string };

export interface SessionManagerOptions {
  factory: BrowserFactory;
  redis: RedisClientLike;
  instanceId: string;
}

export class SessionManager {
  private readonly factory: BrowserFactory;
  private readonly redis: RedisClientLike;
  private readonly instanceId: string;
  private browserPromise: Promise<Browser> | null = null;
  private readonly registered = new Set<string>();
  private readonly sessions = new Map<string, Session>();
  private readonly elements = new ElementRegistry();

  constructor(options: SessionManagerOptions) {
    this.factory = options.factory;
    this.redis = options.redis;
    this.instanceId = options.instanceId;
  }

  /** Returns the adapter_instance_id this manager writes to Redis.
   * Exposed for tests; the value is generated once at startup and
   * never changes. */
  get adapterInstanceId(): string {
    return this.instanceId;
  }

  async register(sessionId: string): Promise<void> {
    const now = new Date().toISOString();
    const metadata: SessionMetadata = {
      session_id: sessionId,
      adapter: ADAPTER_NAME,
      adapter_instance_id: this.instanceId,
      created_at: now,
      last_active_at: now,
      metadata: {},
    };
    // Writes Redis first; if it fails the caller exits Initialize
    // with UNAVAILABLE and the local set is never updated. Order
    // matters: a Redis failure must leave the SessionManager
    // unaware of the id so a retry produces a fresh write rather
    // than the appearance of a registered-but-not-stored session.
    await this.redis.setSession(sessionId, metadata);
    this.registered.add(sessionId);
  }

  has(sessionId: string): boolean {
    return this.registered.has(sessionId);
  }

  async validate(sessionId: string): Promise<ValidationResult> {
    const metadata = await this.redis.getSession(sessionId);
    if (metadata === null) {
      return { kind: "unknown" };
    }
    if (metadata.adapter_instance_id !== this.instanceId) {
      return {
        kind: "different-instance",
        storedInstanceId: metadata.adapter_instance_id,
      };
    }
    // Refresh `last_active_at` and the TTL via a SET-with-EX
    // round-trip. ADR-0023 §5 R4.3 addendum and the phase prompt
    // §4.5 commit to last-write-wins: a second concurrent
    // validate call may overwrite the timestamp by a few ms,
    // which is harmless.
    metadata.last_active_at = new Date().toISOString();
    await this.redis.setSession(sessionId, metadata);
    return { kind: "ok", metadata };
  }

  async getOrCreatePage(sessionId: string): Promise<Page> {
    if (!this.registered.has(sessionId)) {
      throw new UnknownSessionError(sessionId);
    }
    const existing = this.sessions.get(sessionId);
    if (existing) {
      return existing.page;
    }
    const browser = await this.getBrowser();
    const context = await browser.newContext();
    const page = await context.newPage();
    this.sessions.set(sessionId, { context, page });
    return page;
  }

  /** Returns the live `Page` for a session if one exists, or null
   * if the session is registered but has not yet navigated. */
  pageOf(sessionId: string): Page | null {
    return this.sessions.get(sessionId)?.page ?? null;
  }

  bumpGeneration(sessionId: string): void {
    this.elements.bumpGeneration(sessionId);
  }

  currentGeneration(sessionId: string): number {
    return this.elements.currentGeneration(sessionId);
  }

  allocateRefs(sessionId: string, locators: readonly Locator[]): string[] {
    return this.elements.allocateRefs(sessionId, locators);
  }

  lookupRef(sessionId: string, id: string): RefLookup {
    return this.elements.lookupRef(sessionId, id);
  }

  async closeSession(sessionId: string): Promise<boolean> {
    if (!this.registered.has(sessionId)) {
      return false;
    }
    this.registered.delete(sessionId);
    this.elements.forgetSession(sessionId);
    // Best-effort delete; the TTL is the safety net if Redis is
    // briefly unreachable. ADR-0023 §5 / phase prompt §4.6.
    await this.redis.deleteSession(sessionId).catch((err: unknown) => {
      process.stderr.write(
        `redis delete failed for session ${sessionId}: ${String(err)}\n`,
      );
    });
    const session = this.sessions.get(sessionId);
    if (session) {
      this.sessions.delete(sessionId);
      // Closing the context closes its pages; we never close the
      // shared Browser here — other sessions continue.
      await session.context.close().catch(() => undefined);
    }
    return true;
  }

  async closeAll(): Promise<void> {
    const sessions = [...this.sessions.values()];
    this.sessions.clear();
    for (const id of [...this.registered]) {
      this.elements.forgetSession(id);
    }
    this.registered.clear();
    for (const { context } of sessions) {
      await context.close().catch(() => undefined);
    }

    if (this.browserPromise) {
      const browser = await this.browserPromise.catch(() => null);
      this.browserPromise = null;
      if (browser) {
        await browser.close().catch(() => undefined);
      }
    }
  }

  private getBrowser(): Promise<Browser> {
    if (!this.browserPromise) {
      this.browserPromise = this.factory();
    }
    return this.browserPromise;
  }
}

File: adapters/seleniumbase/src/spectre_seleniumbase/sessions.py
"""Session manager for the SeleniumBase adapter.

Owns the lazy WebDriver launch, the per-session driver
allocation, and the per-session :class:`ElementRegistry`
that backs ``Query``, ``Extract``, and the element-scoped
``Screenshot``. R4.3 externalises session metadata to Redis
(ADR-0023 §4 + §5): ``register`` writes the metadata document,
``validate`` reads it on every non-Initialize RPC, and
``close_session`` deletes the key (best-effort). The runtime
WebDriver objects stay process-local — the §5 restart-invalidation
contract makes Pod restart equivalent to session loss for clients.

Contract (R4.3):

- ``register(session_id)`` — ``Initialize`` calls this. Writes a
  :class:`SessionMetadata` document to Redis with the current
  ``adapter_instance_id`` and adds the id to the local set.
  Raises if Redis is unreachable; the caller maps the failure to
  gRPC ``UNAVAILABLE``.
- ``validate(session_id)`` — every non-Initialize RPC calls this
  before doing any work. Returns one of three :class:`Validation`
  kinds (``ok`` / ``unknown`` / ``different_instance``). The
  ``ok`` path refreshes ``last_active_at`` and the TTL (last-
  write-wins per phase prompt §4.5).
- ``has(session_id)`` — true once ``register`` has been called for
  the id locally and ``close_session`` / ``close_all`` have not
  yet evicted it. Used for cheap membership checks where Redis is
  not authoritative.
- ``get_or_create_driver(session_id)`` — first call for a
  registered id launches the per-session WebDriver. Subsequent
  calls with the same id return the same driver. Raises
  :class:`UnknownSessionError` if the id is not registered.
- ``driver_of(session_id)`` — returns the launched driver or
  ``None`` (no UnknownSessionError; used by handlers that need
  to distinguish "session unknown" from "session known but
  Navigate not yet called").
- ``bump_generation`` / ``current_generation`` /
  ``register_element`` / ``register_elements`` /
  ``lookup_element`` — element-registry delegation per ADR-0010.
- ``close_session(session_id)`` — quits the WebDriver if one was
  launched, evicts the local entry, forgets the ElementRegistry
  entry, and best-effort deletes the Redis key (TTL is the
  safety net; phase prompt §4.6).
- ``close_all()`` — quits every driver, clears every local
  registration, forgets every registry entry. Idempotent.
  Does not enumerate Redis keys for deletion — restart
  invalidation handles abandoned keys via TTL expiry.

The class is constructed with a ``DriverFactory``, a
``RedisClient``, and the adapter's ``instance_id``. Unit tests
pass a ``fakeredis``-backed client and any ``instance_id``
string.

See ADR-0014, ADR-0015, ADR-0023 §4–§5 (R4.3), and ADR-0009 for
the lifecycle rationale.
"""

from __future__ import annotations

import contextlib
import sys
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Literal

from .elements import ElementRegistry, RefLookup
from .redis_client import ADAPTER_NAME, RedisClient, SessionMetadata

DriverFactory = Callable[[], Any]


class UnknownSessionError(RuntimeError):
    """Raised when an RPC references a ``session_id`` that ``Initialize`` did
    not produce."""

    def __init__(self, session_id: str) -> None:
        super().__init__(f"unknown session_id: {session_id!r}")
        self.session_id = session_id


@dataclass
class Validation:
    """Result of :meth:`SessionManager.validate`.

    ``kind`` is ``"ok"`` when Redis has the session and the
    stored ``adapter_instance_id`` matches, ``"unknown"`` when
    Redis has no entry, and ``"different_instance"`` when the
    stored id belongs to a foreign adapter instance — the §5
    restart-invalidation case the gRPC handler maps to
    ``UNAVAILABLE``.
    """

    kind: Literal["ok", "unknown", "different_instance"]
    metadata: SessionMetadata | None = None
    stored_instance_id: str | None = None


@dataclass
class _Session:
    driver: Any | None = None


@dataclass
class SessionManager:
    """In-memory registry mapping ``session_id`` → WebDriver instance,
    backed by Redis-resident metadata for the §5 restart-invalidation
    contract.

    The Playwright adapter's ``SessionManager`` (TypeScript) and this
    one serve the same role; ADR-0014 records the deliberate decision
    to re-implement the shape rather than extract a shared contract
    before a third driver lands. ADR-0015 §1 carries the same
    reasoning forward to the ElementRegistry integration; ADR-0023 §8
    extends it to per-language Redis libraries.
    """

    factory: DriverFactory
    redis: RedisClient
    instance_id: str
    _sessions: dict[str, _Session] = field(default_factory=dict, init=False, repr=False)
    _elements: ElementRegistry = field(default_factory=ElementRegistry, init=False, repr=False)

    @property
    def adapter_instance_id(self) -> str:
        """Expose the adapter's instance UUID for tests and diagnostics."""
        return self.instance_id

    def register(self, session_id: str) -> None:
        """Write the session metadata to Redis and register the id locally.

        Order matters: a Redis write failure must leave the local
        set unaware of the id so a retry produces a fresh write
        rather than the appearance of a registered-but-not-stored
        session.
        """
        now = _utc_now_iso()
        metadata: SessionMetadata = {
            "session_id": session_id,
            "adapter": ADAPTER_NAME,
            "adapter_instance_id": self.instance_id,
            "created_at": now,
            "last_active_at": now,
            "metadata": {},
        }
        self.redis.set_session(session_id, metadata)
        self._sessions.setdefault(session_id, _Session())

    def has(self, session_id: str) -> bool:
        """Return ``True`` if the session is registered locally."""
        return session_id in self._sessions

    def validate(self, session_id: str) -> Validation:
        """Validate the session against Redis.

        Returns a :class:`Validation` whose ``kind`` distinguishes
        ``ok`` / ``unknown`` / ``different_instance``. The ``ok``
        path refreshes ``last_active_at`` and the TTL via a SET
        round-trip — last-write-wins per phase prompt §4.5.
        """
        metadata = self.redis.get_session(session_id)
        if metadata is None:
            return Validation(kind="unknown")
        stored_id = metadata.get("adapter_instance_id", "")
        if stored_id != self.instance_id:
            return Validation(kind="different_instance", stored_instance_id=stored_id)
        metadata["last_active_at"] = _utc_now_iso()
        self.redis.set_session(session_id, metadata)
        return Validation(kind="ok", metadata=metadata)

    def get_or_create_driver(self, session_id: str) -> Any:
        """Return the session's driver, launching it on first call.

        Raises :class:`UnknownSessionError` if the id is not registered.
        """
        session = self._sessions.get(session_id)
        if session is None:
            raise UnknownSessionError(session_id)
        if session.driver is None:
            session.driver = self.factory()
        return session.driver

    def driver_of(self, session_id: str) -> Any | None:
        """Return the session's driver if one has been launched, else ``None``."""
        session = self._sessions.get(session_id)
        return session.driver if session is not None else None

    # -- ElementRegistry delegation -------------------------------------

    def bump_generation(self, session_id: str) -> None:
        """Invalidate every prior ElementRef for the session.

        Called after a successful ``Navigate``. See ADR-0010 §1
        and ADR-0015 §1.
        """
        self._elements.bump_generation(session_id)

    def current_generation(self, session_id: str) -> int:
        """Return the session's current generation counter, or zero."""
        return self._elements.current_generation(session_id)

    def register_element(self, session_id: str, web_element: Any) -> str:
        """Allocate a UUID for a single ``WebElement``."""
        ids = self._elements.allocate_refs(session_id, [web_element])
        return ids[0]

    def register_elements(self, session_id: str, web_elements: list[Any]) -> list[str]:
        """Allocate UUIDs for a list of ``WebElement``s in order."""
        return self._elements.allocate_refs(session_id, list(web_elements))

    def lookup_element(self, session_id: str, ref_id: str) -> RefLookup:
        """Resolve a UUID back to its ``WebElement``."""
        return self._elements.lookup_ref(session_id, ref_id)

    # -- Lifecycle ------------------------------------------------------

    def close_session(self, session_id: str) -> bool:
        """Tear down the local driver, forget the registry entry, and
        best-effort delete the Redis key.

        Returns ``True`` if the session was registered locally,
        ``False`` if the id was unknown. Errors raised by
        ``driver.quit()`` and by Redis ``DEL`` are swallowed —
        the session is forgotten regardless. The Redis TTL is the
        safety net for the rare delete failure (§4.6).
        """
        session = self._sessions.pop(session_id, None)
        if session is None:
            return False
        self._elements.forget_session(session_id)
        try:
            self.redis.delete_session(session_id)
        except Exception as exc:  # noqa: BLE001 — best-effort delete
            sys.stderr.write(
                f"redis delete failed for session {session_id}: {exc}\n",
            )
            sys.stderr.flush()
        if session.driver is not None:
            with contextlib.suppress(Exception):
                session.driver.quit()
        return True

    def close_all(self) -> None:
        """Quit every driver and clear every registration. Idempotent."""
        session_ids = list(self._sessions.keys())
        sessions = list(self._sessions.values())
        self._sessions.clear()
        for session_id in session_ids:
            self._elements.forget_session(session_id)
        for session in sessions:
            if session.driver is not None:
                with contextlib.suppress(Exception):
                    session.driver.quit()


def _utc_now_iso() -> str:
    """Return the current UTC time formatted as ISO-8601 with Z suffix.

    Matches the Playwright adapter's ``new Date().toISOString()``
    output (``YYYY-MM-DDTHH:MM:SS.sssZ``) so the JSON document
    round-trips identically through any adapter.
    """
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"
